server/game/ChessClock.js

```diff
@@ -0,0 +1,98 @@
+class ChessClock {
+    constructor(player, time, delayToStartClock) {
+        this.player = player;
+        this.mainTime = time * 60;
+        this.timeLeft = time * 60;
+        this.mode = 'inactive';
+        this.timerStart = 0; //when the clock is running, this contains the instant on which it started
+        this.paused = false; //has the clock been paused via chat command?
+        this.running = false; //is the clock ticking for the player?
+        this.stateId = 0;
+        this.name = 'ChessClock';
+        this.activated = false; //indicates if the game has started/setup is finished 
+        this.delayToStartClock = delayToStartClock;
+
+        this.player.game.on('onSetupFinished', () => this.activateChessClock());
+    }
+
+    activateChessClock() {
+        this.activated = true;
+    }
+
+    togglePause() {
+        if(this.activated) {
+            this.paused = !this.paused;
+            if(this.paused) {
+                this.stop();
+            }
+        }
+    }
+
+    modify(secs) {
+        this.timeLeft += secs;
+    }
+
+    updateStateId() {
+        this.stateId++;
+    }
+
+    start() {
+        if(this.activated) {
+            if(!this.paused && !this.running) {
+                this.mode = 'down';
+                this.running = true;
+                this.timerStart = Date.now();
+                this.updateStateId();
+            }
+        }
+    }
+
+    stop() {
+        if(this.activated) {
+            if(this.timerStart > 0 && this.running) {
+                this.running = false;
+                this.updateTimeLeft(Math.floor(((Date.now() - this.timerStart) / 1000) + 0.5));
+                this.timerStart = 0;
+                this.updateStateId();
+            }
+            this.mode = 'stop';
+        }
+    }
+
+    timeRanOut() {
+        this.player.game.addMessage('{0}\'s clock has run out', this.player);
+        this.player.game.recordWinner(this.player.game.getOpponents(this.player)[0], 'time');

```

Comment:
This is not melee friendly. I don't think we need to make it so right now, but perhaps leave a TODO comment so we don't lose track of it?

Reply:
Good idea, TODO added